File: algorithms/medium/find_duplicate_file_in_system.ex
defmodule Solution do
  @spec find_duplicate(paths :: [String.t()]) :: [[String.t()]]
  def find_duplicate(paths) do
    paths
    |> Enum.reduce(%{}, &aux/2)
    |> Map.values()
    |> Enum.filter(&(Enum.count(&1) > 1))
  end

  def aux(path, m) do
    [dirname | files] = String.split(path)

    files
    |> Enum.map(&String.split(&1, ["(", ")"]))
    |> Enum.reduce(m, fn [filename, content | _rest], acc ->
      name = dirname <> "/" <> filename
      Map.update(acc, content, [name], &[name | &1])
    end)
  end
end

File: algorithms/medium/find_duplicate_file_in_system.rs
use std::collections::HashMap;

fn aux(s: &str, map: &mut HashMap<String, Vec<String>>) {
    let mut iter = s.split_whitespace();
    let dirname = iter.next().unwrap();
    for file in iter {
        let mut name_content_iter = file.split(['(', ')'].as_ref());
        let filepath = format!("{}/{}", dirname, name_content_iter.next().unwrap());
        let content = name_content_iter.next().unwrap().to_string();
        (*map.entry(content).or_insert(vec![])).push(filepath);
    }
}

impl Solution {
    pub fn find_duplicate(paths: Vec<String>) -> Vec<Vec<String>> {
        let mut map = HashMap::new();
        for p in &paths {
            aux(p, &mut map);
        }

        map.values()
            .filter_map(|v| if v.len() < 2 { None } else { Some(v.to_vec()) })
            .collect()
    }
}
